File: Back-End/Db/models/task_model.js
import mongoose, { model, Schema, Types } from "mongoose";
import { queryHelperPlugin } from "../query_helpers.js";

const taskSchema = new Schema(
  {
    title: {
      type: String,
      required: true,
      trim: true,
      minlength: [3, "Title must be at least 3 characters"],
      maxlength: [100, "Title must be at most 100 characters"],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [500, "Description must be at most 500 characters"],
    },
    createdBy: {
      type: Types.ObjectId,
      required: true,
      ref: "Parent",
    },

    assignedTo: {
      type: Types.ObjectId,
      ref: "Child",
      default: null,
      required: function () {
        return this.type === "personal";
      },
    },

    type: {
      type: String,
      enum: ["personal", "open"],
      required: true,
    },

    claimedBy: {
      type: Types.ObjectId,
      ref: "Child",
      default: null,
    },

    points: {
      type: Number,
      min: [1, "Points cannot be negative"],
      default: 1,
    },
    priority: {
      type: String,
      enum: ["low", "medium", "high"],
      default: "medium",
    },
    category: {
      type: String,
      required: true,
      trim: true,
      lowercase: true,
    },
    status: {
      type: String,
      enum: ["pending", "claimed", "rejected", "submitted", "approved"],
      default: "pending",
    },
    image: {
      url: {
        type: String,
      },
      id: {
        type: String,
      },
    },

    requirements: {
      submissionType: {
        type: String,
        enum: ["text", "image", "both"],
        default: "text",
      },
      minImages: {
        type: Number,
        default: 0,
        min: 0,
      },
    },

    submission: {
      description: {
        type: String,
        maxlength: [500, "Submission description too long"],
      },
      images: [
        {
          url: String,
          id: String,
        },
      ],
      submittedAt: Date,
    },
    isDeleted: {
      type: Boolean,
      default: false,
    },
    dueDate: Date,
    completedAt: Date,
    approvedAt: Date,
    rejectionReason: String,
  },
  { timestamps: true, toJSON: true, toObject: true },
);

taskSchema.virtual("isOverdue").get(function () {
  return (
    this.dueDate && this.status !== "approved" && this.dueDate < new Date()
  );
});

taskSchema.virtual("isClaimed").get(function () {
  return !!this.claimedBy;
});

//return clean res and remove empty fields
taskSchema.set("toJSON", {
  transform: (doc, ret) => {
    ret.taskId = ret._id;
    delete ret._id;
    delete ret.__v;

    // remove empty claimedBy
    if (!ret.claimedBy) delete ret.claimedBy;

    //remove assignedTo if null
    if (!ret.assignedTo) delete ret.assignedTo;
    // remove empty submission
    if (!ret.submission?.images?.length && !ret.submission?.description) {
      delete ret.submission;
    }
    // remove empty image
    if (!ret.image?.url) {
      delete ret.image;
    }

    return ret;
  },
});

taskSchema.index({ assignedTo: 1, status: 1, dueDate: 1 });
taskSchema.index({ createdBy: 1, status: 1, createdAt: -1 });

taskSchema.plugin(queryHelperPlugin);

const Task = mongoose.models.Task || model("Task", taskSchema);

export default Task;

File: Front-End/lib/features/tasks/data/models/task_model.dart
/// Task model for `GET task/tasks` (data[]) and `GET task/:id` (task).
class TaskModel {
  final String taskId;
  final String title;
  final String description;
  final int points;
  final String priority;
  final String category;
  final String type; // personal | open
  final String status; // pending | submitted | approved | rejected | ...
  final String assignedTo;
  final String createdBy;
  final String createdAt;
  final String dueDate;

  final String submissionType; // image
  final int minImages;

  final List<String> submissionImages;
  final String submittedAt;
  final String aiReviewStatus;
  final String aiReviewReasoning;

  final String taskImageUrl;

  const TaskModel({
    required this.taskId,
    required this.title,
    required this.description,
    required this.points,
    required this.priority,
    required this.category,
    required this.type,
    required this.status,
    required this.assignedTo,
    required this.createdBy,
    required this.createdAt,
    required this.dueDate,
    required this.submissionType,
    required this.minImages,
    required this.submissionImages,
    required this.submittedAt,
    required this.aiReviewStatus,
    required this.aiReviewReasoning,
    required this.taskImageUrl,
  });

  bool get isOpen => type == 'open';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
  bool get isSubmitted => status == 'submitted';
  bool get needsImages => submissionType == 'image';

  static int _toInt(dynamic v) =>
      v is num ? v.toInt() : (int.tryParse('$v') ?? 0);

  factory TaskModel.fromJson(Map<String, dynamic> json) {
    final req = (json['requirements'] as Map?)?.cast<String, dynamic>() ?? const {};
    final sub = (json['submission'] as Map?)?.cast<String, dynamic>();
    final taskImg = json['taskImage'] ?? json['taskImg'];

    final List<String> images = [];
    if (sub != null && sub['images'] is List) {
      for (final img in (sub['images'] as List)) {
        if (img is Map && img['url'] != null) images.add(img['url'].toString());
      }
    }
    final aiReview = sub != null ? sub['aiReview'] : null;
    final aiResult = aiReview is Map ? aiReview['result'] : null;

    return TaskModel(
      taskId: (json['taskId'] ?? json['_id'] ?? '').toString(),
      title: (json['title'] ?? '').toString(),
      description: (json['description'] ?? '').toString(),
      points: _toInt(json['points']),
      priority: (json['priority'] ?? '').toString(),
      category: (json['category'] ?? '').toString(),
      type: (json['type'] ?? '').toString(),
      status: (json['status'] ?? '').toString(),
      assignedTo: (json['assignedTo'] ?? '').toString(),
      createdBy: (json['createdBy'] ?? '').toString(),
      createdAt: (json['createdAt'] ?? '').toString(),
      dueDate: (json['dueDate'] ?? '').toString(),
      submissionType: (req['submissionType'] ?? '').toString(),
      minImages: _toInt(req['minImages']),
      submissionImages: images,
      submittedAt: sub != null ? (sub['submittedAt'] ?? '').toString() : '',
      aiReviewStatus:
          aiReview is Map ? (aiReview['status'] ?? '').toString() : '',
      // The reasoning text lives at `aiReview.reason` (older data used
      // `aiReview.result.reasoning`).
      aiReviewReasoning: aiReview is Map
          ? (aiReview['reason'] ??
                  (aiResult is Map ? aiResult['reasoning'] : null) ??
                  '')
              .toString()
          : '',
      taskImageUrl:
          taskImg is Map ? (taskImg['url']?.toString() ?? '') : '',
    );
  }
}
